### ISSUE TITLE
bug: grpc-web plugin causing missing trailers issue

### ISSUE THREAD

**Issue by Affanmir** [apache/apisix#10573]
### Current Behavior

When i use the grpc-web plugin and cors plugin to communicate from my fronted to my backend, the response I receive an error missing trailers. I've tried the same thing using envoy and it works. There seems to be some misconfiguration is the response body which does not allow the response to

### Expected Behavior
![Screenshot 2023-11-28 at 10 52 01 AM](https://github.com/apache/apisix/assets/54717435/cb92ca3f-9669-49e0-8b3f-cd4d5d7dcc95)

Envoy Response: (CORRECT AND DOES NOT CAUSE ISSUE)

Hello! Mohsin Javed� grpc-status:0
grpc-message:OK

Headers:

HTTP/1.1 200 OK
content-type: application/grpc-web+proto
grpc-accept-encoding: identity,deflate,gzip
grpc-encoding: identity
date: Fri, 24 Nov 2023 06:04:45 GMT
x-envoy-upstream-service-time: 12
access-control-allow-origin: http://localhost:58653/
access-control-expose-headers: custom-header-1,grpc-status,grpc-message
server: envoy
transfer-encoding: chunked

API SIX Response:: (INCORRECT 

AND CAUSES ISSUE MISSING TRAILERS)

Hello! World

Headers:
HTTP/1.1 200 OK
Date: Fri, 24 Nov 2023 06:13:48 GMT
Content-Type: application/grpc-web+proto
Transfer-Encoding: chunked
Connection: keep-alive
grpc-accept-encoding: identity,deflate,gzip
grpc-encoding: identity
Server: APISIX/3.6.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: *
Access-Control-Max-Age: 5
Access-Control-Expose-Headers: *
Access-Control-Allow-Headers: *

### Error Logs

<img width="538" alt="Screenshot 2023-11-30 at 7 05 22 PM" src="https://github.com/apache/apisix/assets/54717435/8c10efce-295a-47a8-9315-aae70d5935c4">

### Steps to Reproduce

enable grpc-web and cors plugin on a route that communicates from any frontend / backend service that implemented a grpc service using a proto definition.

### Environment

Please take a clone of this https://github.com/m-jayy/grpc_with_envoy_and_apisix/pull/1 All steps to reproduce are mentioned in the README.md

This is a small local setup with node as the back end and flutter as the frontend

**Comment by shreemaan-abhishek** [comment]
could you please provide me with reproduction steps and reproduction environment identical to yours? This will save us a lot of back and forth.

**Comment by Affanmir** [comment]
@shreemaan-abhishek Please take a clone of this [PR](https://github.com/m-jayy/grpc_with_envoy_and_apisix/pull/1)  All steps to reproduce are mentioned in the README.md

This is a small local setup with node as the back end and flutter as the frontend

**Comment by kayx23** [comment]
same issue: https://github.com/apache/apisix/issues/10254

**Comment by shreemaan-abhishek** [comment]
@Affanmir I'm taking a look at this, mean while could you please share your route and upstream configuration? I want to confirm if you have used the `grpc` scheme correctly: 
<img width="763" alt="image" src="https://github.com/apache/apisix/assets/61597896/e34fc2f6-daae-450f-9f3f-cee69c563b00">

**Comment by Affanmir** [comment]
<img width="760" alt="Screenshot 2023-12-04 at 11 18 53 AM" src="https://github.com/apache/apisix/assets/54717435/62891317-5070-4cfe-a9b8-1f13e05f17c7">
<img width="1103" alt="Screenshot 2023-12-04 at 11 19 05 AM" src="https://github.com/apache/apisix/assets/54717435/5280e868-c8ed-4780-a440-0944821084ad">
<img width="343" alt="Screenshot 2023-12-04 at 11 19 21 AM" src="https://github.com/apache/apisix/assets/54717435/f101d354-f1b2-492b-bd6b-52c31e92aff9">
<img width="333" alt="Screenshot 2023-12-04 at 11 19 32 AM" src="https://github.com/apache/apisix/assets/54717435/7504d315-7774-4a7a-b69f-f6f37e4e1965">

**Comment by alifahadk** [comment]
@shreemaan-abhishek We are trying to use this plugin for proxying between a gRPC web client (more specifically, the grpc dart web client) and a grpc-java server.

Our grpc-java server does not return a grpc-web-compliant HTTP/1.1 response on its own (with the trailers present in the response body). Is this plugin meant to cater to such a scenario (similar to the support provided by the [grpc web filter by Envoy](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_web_filter))? Our initial assumption was that it does that, but latest explorations suggest otherwise.

**Comment by Affanmir** [comment]
@shreemaan-abhishek 
`looks like trailers as part of response body is also not being supported.`
Yes exactly on point. Is this by design or a bug?

**Comment by shreemaan-abhishek** [comment]
Not a bug, it's just not supported. You may take this up if you want to fix this.

**Comment by Affanmir** [comment]
@shreemaan-abhishek Is it possible for you to provide us some resources on where this functionality needs to be added? are there any conventions or guidelines we should be following? Essentially what I'm thinking is to add this logic in the plugin itself. But it will require some changes to response body and exposing some additional headers as well.

**Comment by sheharyaar** [comment]
@Affanmir , is the associated PR ready or can I take up this issue and work on a PR ? 
Your PR implements the various CORS requirements as mentioned in the grp-c doc : https://github.com/grpc/grpc-web/blob/master/doc/browser-features.md#cors-support

We will need to check and implement the entire requirement doc.

**Comment by Affanmir** [comment]
@sheharyaar it is missing, absolute pathing some additional DEFAULT_CORS_ALLOW_HEADERS, the missing trailers issue is fixed in the associated PR but these extra functionality will need to be incorporated 
 
Current behavior of grpc is following 
IMPORTANT
While using the grpc-web Plugin, always use a prefix matching pattern (/*, /grpc/example/*) for matching Routes. This is because the gRPC Web client passes the package name, the service interface name, the method name and other information in the proto in the URI. For example, /path/a6.RouteService/Insert.

Ive made it work locally and added parameters to enable/disable will be updating the PR soon this week

**Comment by sheharyaar** [comment]
Cool, we also need to have HTTP to GRPC mapping : https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md

> This table is to be used only for clients that received a response that did not include grpc-status. If grpc-status was provided, it must be used. Servers must not use this table to determine an HTTP status code to use; the mappings are neither symmetric nor 1-to-1.

**Comment by Affanmir** [comment]
@sheharyaar maybe you can create a separate PR for that mapping

**Comment by sheharyaar** [comment]
Ok, once your PR is merged I can do that in another PR.

**Comment by sheharyaar** [comment]
@Affanmir , I will be taking up this PR, I will fix the CI and do the next steps.

**Comment by sheharyaar** [comment]
@shreemaan-abhishek can you please assign me this ?

**Comment by sheharyaar** [comment]
Sure, just please list out the tasks that are pending. I will take up the rest and once the PR is ready, I will ping you.

**Comment by sheharyaar** [comment]
@Affanmir , I noticed that when you use grpc-web-text, the reponse _has_ the trailers (grpc-status and grpc-message)

```bash
curl -i 'http://127.0.0.1:9080/grpc/web/a6.RouteService/GetRoute' \
                      -H 'Accept: application/grpc-web-text' \
                      -H 'Accept-Language: en-IN,en-US;q=0.9,en-GB;q=0.8,en;q=0.7' \
                      -H 'Connection: keep-alive' \
                      -H 'Content-Type: application/grpc-web-text' \
                      --data-raw 'AAAAAAcKBWhlbGxv' \
                      
HTTP/1.1 200 OK
Date: Wed, 17 Jan 2024 11:30:05 GMT
Content-Type: application/grpc-web-text
Transfer-Encoding: chunked
Connection: keep-alive
Server: APISIX/3.8.0
Vary: Origin
Access-Control-Allow-Origin: *

AAAAAA8KBWhlbGxvEgYvaGVsbG8=grpc-status: 0
grpc-message: 
```
